resolved: Fix DoT timeout on multiple answer records (for CloudFlare, Google, etc. DoT servers)

joanbm opened the pull request:
When sending multiple DNS questions to a DNS-over-TLS server (e.g. a question for A and AAAA records, as is typical) on the same session, the server may answer to each question in a separate TLS record, but it may also aggregate multiple answers in a single TLS record.
(Some servers do this very often (e.g. Cloudflare 1.0.0.1), some do it sometimes (e.g. Google 8.8.8.8) and some seem to never do it (e.g. Quad9 9.9.9.10)).

Both cases should be handled equivalently, as the byte stream is the same, but when multiple answers came in a single TLS record, usually the first answer was processed, but the second answer was entirely ignored, which caused a 10s delay until the resolution timed out and the missing question was retried. This can be reproduced by configuring one of the offending server and running `resolvectl query google.com --cache=no` a few times.

To be notified of incoming data, systemd-resolved listens to `EPOLLIN` events on the underlying socket. However, when DNS-over-TLS is used, the TLS library (OpenSSL or GnuTLS) may read and buffer the entire TLS record when reading the first answer, so usually no further `EPOLLIN` events will be generated, and the second answer will never be processed.

To avoid this, if there's buffered TLS data, generate a "fake" EPOLLIN event. This is hacky, but it makes this case transparent to the rest of the IO code.

Fixes: https://bbs.archlinux.org/viewtopic.php?id=262926

--

**Appendix:**

To verify the issue, you can use socat like so to intercept some DoT server: `socat -v -x openssl-listen:853,reuseaddr,fork,cert=127.0.0.1.pem,key=127.0.0.1-key.pem,verify=0 openssl:8.8.8.8:853` (you will need a trusted certificate for localhost).

Sample stream that did not cause systemd-resolved to hang:
```
> 2022/01/15 22:23:48.849930  length=2 from=820 to=821
 00 27                                            .'
--
> 2022/01/15 22:23:48.850041  length=39 from=822 to=860
 f2 43 01 00 00 01 00 00 00 00 00 01 06 67 6f 6f  .C...........goo
 67 6c 65 03 63 6f 6d 00 00 01 00 01 00 00 29 ff  gle.com.......).
 e4 00 00 00 00 00 00                             .......
--
> 2022/01/15 22:23:48.850178  length=2 from=861 to=862
 00 27                                            .'
--
> 2022/01/15 22:23:48.850230  length=39 from=863 to=901
 b0 30 01 00 00 01 00 00 00 00 00 01 06 67 6f 6f  .0...........goo
 67 6c 65 03 63 6f 6d 00 00 1c 00 01 00 00 29 ff  gle.com.......).
 e4 00 00 00 00 00 00                             .......
--
< 2022/01/15 22:23:48.883489  length=69 from=1260 to=1328
 00 43 b0 30 81 80 00 01 00 01 00 00 00 01 06 67  .C.0...........g
 6f 6f 67 6c 65 03 63 6f 6d 00 00 1c 00 01 c0 0c  oogle.com.......
 00 1c 00 01 00 00 00 87 00 10 2a 00 14 50 40 03  ..........*..P@.
 08 06 00 00 00 00 00 00 20 0e 00 00 29 02 00 00  ........ ...)...
 00 00 00 00 00                                   .....
--
< 2022/01/15 22:23:48.883906  length=57 from=1329 to=1385
 00 37 f2 43 81 80 00 01 00 01 00 00 00 01 06 67  .7.C...........g
 6f 6f 67 6c 65 03 63 6f 6d 00 00 01 00 01 c0 0c  oogle.com.......
 00 01 00 01 00 00 01 12 00 04 8e fa b2 ae 00 00  ................
 29 02 00 00 00 00 00 00 00                       )........
```

Sample stream that caused systemd-resolved to hang:
```
> 2022/01/15 22:23:50.113990  length=2 from=902 to=903
 00 27                                            .'
--
> 2022/01/15 22:23:50.114074  length=39 from=904 to=942
 d4 53 01 00 00 01 00 00 00 00 00 01 06 67 6f 6f  .S...........goo
 67 6c 65 03 63 6f 6d 00 00 01 00 01 00 00 29 ff  gle.com.......).
 e4 00 00 00 00 00 00                             .......
--
> 2022/01/15 22:23:50.114174  length=2 from=943 to=944
 00 27                                            .'
--
> 2022/01/15 22:23:50.114216  length=39 from=945 to=983
 35 79 01 00 00 01 00 00 00 00 00 01 06 67 6f 6f  5y...........goo
 67 6c 65 03 63 6f 6d 00 00 1c 00 01 00 00 29 ff  gle.com.......).
 e4 00 00 00 00 00 00                             .......
--
< 2022/01/15 22:23:50.147560  length=126 from=1386 to=1511
 00 37 d4 53 81 80 00 01 00 01 00 00 00 01 06 67  .7.S...........g
 6f 6f 67 6c 65 03 63 6f 6d 00 00 01 00 01 c0 0c  oogle.com.......
 00 01 00 01 00 00 01 10 00 04 8e fa b2 ae 00 00  ................
 29 02 00 00 00 00 00 00 00 00 43 35 79 81 80 00  ).........C5y...
 01 00 01 00 00 00 01 06 67 6f 6f 67 6c 65 03 63  ........google.c
 6f 6d 00 00 1c 00 01 c0 0c 00 1c 00 01 00 00 00  om..............
 85 00 10 2a 00 14 50 40 03 08 06 00 00 00 00 00  ...*..P@........
 00 20 0e 00 00 29 02 00 00 00 00 00 00 00        . ...)........
```

joanbm commented:
Update: Fixed error handling for `sd_event_source_get_io_events`.

yuwata commented on src/resolve/resolved-dnstls-gnutls.c:
```diff
@@ -211,6 +211,14 @@ ssize_t dnstls_stream_read(DnsStream *stream, void *buf, size_t count) {
         return ss;
 }
 
+int dnstls_stream_has_buffered_data(DnsStream *stream) {
```
`int` -> `bool`

joanbm replied:
Done.

yuwata commented on src/resolve/resolved-dnstls-openssl.c:
```diff
@@ -367,6 +367,14 @@ ssize_t dnstls_stream_read(DnsStream *stream, void *buf, size_t count) {
         return ss;
 }
 
+int dnstls_stream_has_buffered_data(DnsStream *stream) {
```
`int` -> `bool`

joanbm replied:
Done.

yuwata commented on src/resolve/resolved-dnstls-openssl.c:
```diff
@@ -367,6 +367,14 @@ ssize_t dnstls_stream_read(DnsStream *stream, void *buf, size_t count) {
+        assert(stream->encrypted);
+        assert(stream->dnstls_data.ssl);
+
+        return SSL_has_pending(stream->dnstls_data.ssl);
```
The man page says it returns 0 or 1, but for safety `return SSL_has_pending() > 0;`

joanbm replied:
Done.

yuwata commented on src/resolve/resolved-dns-stream.c:
```diff
@@ -441,6 +453,34 @@ static int on_stream_io(sd_event_source *es, int fd, uint32_t revents, void *use
+                   This is hacky, but it makes this case transparent to the rest of the IO code. */
+                if ((events & EPOLLIN) != 0 && dnstls_stream_has_buffered_data(s)) {
+                        if (!s->tls_buffered_data_event_source) {
+                                r = sd_event_add_defer(s->manager->event, &s->tls_buffered_data_event_source,
```
Why do not you simply loop this function?

joanbm replied:
Looping is possible, it's more of an style thing:
(1) Besides the declarations the entire handling of the edge case is contained in a single block
(2) Looping/recursing/goto'ing to the top of the function feels ugly (more indentation or stack frames or labels, inside #ifdefs for ENABLE_DNS_OVER_TLS)
(3) The code flow is more unified with the plain TCP case, e.g. an event emits at most one DNS packet

But I understand that emitting a "fake IO" event is also a bit ugly and less performant, so if you feel looping is preferable, I can change that.

yuwata replied:
If looping also works, then I prefer to convert the logic you added into loop.

How about to rename `on_stream_io()` to e.g. `on_stream_io_impl()` and then
```
static int on_stream_io_impl(DnsStream *s, uint32_t revents) {
        ...
}

static int on_stream_io(sd_event_source *es, int fd, uint32_t revents, void *userdata) {
        _cleanup_(dns_stream_unrefp) DnsStream *s = NULL;
        int r;

        assert(userdata);

        s = dns_stream_ref(userdata); /* Protect stream while we process it */

        r = on_stream_io(s, revents);
        if (r < 0)
                return r;

#if ENABLE_DNS_OVER_TLS
        if (!s->encrypted)
                return 0;
        if (!FLAGS_SET(events, EPOLLIN))
                return 0;

        while (dnstls_stream_has_buffered_data(s)) {
                /* When using DNS-over-TLS, the underlying TLS library may read the entire TLS record
                   and buffer it internally. If this happens, we will not receive further EPOLLIN events,
                   and unless there's some unrelated activity on the socket, we will hang until time out.
                   To avoid this, ....... */
                r = on_stream_io_impl(s, revents);
                if (r < 0)
                        return r;
        }
#endif

        return 0;
}
```

yuwata commented on src/resolve/resolved-dns-stream.c:
```diff
@@ -441,6 +453,34 @@ static int on_stream_io(sd_event_source *es, int fd, uint32_t revents, void *use
+                   and unless there's some unrelated activity on the socket, we will hang until time out.
+                   To avoid this, if there's buffered TLS data, generate a "fake" EPOLLIN event.
+                   This is hacky, but it makes this case transparent to the rest of the IO code. */
+                if ((events & EPOLLIN) != 0 && dnstls_stream_has_buffered_data(s)) {
```
Not a matter, but I prefer `FLAGS_SET(events, EPOLLIN)`.

joanbm replied:
Agreed and done.

yuwata reviewed (commented):
Sorry for late review.

joanbm commented:
Reviewed according to comments + added an automated test for DnsStream reproducing the bug.

yuwata reviewed (commented):
Thank you for adding test. That's super nice!

joanbm commented:
Ready for re-review:
* Converted logic to handle TLS buffered data from defer event to loop
* Fixed style issues in the tests
* Fix CI: Skip the unit test if the 'openssl' command is not available at runtime. @mrc0mmand not sure if you have any better way to go about this. I don't like to use OpenSSL or GnuTLS as a library in the test, I prefer something like openssl/socat/stunnel/etc. since it's simpler, and harder to get wrong.
* Fix a minor memory safety (ASAN) issue in the test

PS: Sorry for launching so many CI runs, I kept finding small things to fix.

yuwata commented on src/resolve/test-resolved-stream.c:
```diff
@@ -0,0 +1,345 @@
+        test_dns_stream(false);
+#if ENABLE_DNS_OVER_TLS
+        if (system("openssl version >/dev/null 2>&1") != 0) {
+                log_warning("Skipping TLS test since the 'openssl' command does not seem to be available");
```
`return log_tests_skipped("...");` can be used.

joanbm replied:
Good catch, this is much cleaner.

yuwata commented on src/resolve/resolved-dns-stream.c:
```diff
@@ -280,13 +281,10 @@ static int on_stream_timeout(sd_event_source *es, usec_t usec, void *userdata) {
         bool progressed = false;
         int r;
 
-        assert(s);
```
Please do not remove it. Of course, it is non-NULL now, but just for readability.

yuwata commented on src/resolve/resolved-dns-stream.c:
```diff
@@ -441,6 +439,45 @@ static int on_stream_io(sd_event_source *es, int fd, uint32_t revents, void *use
+
+        r = on_stream_io_impl(s, revents);
+        if (r <= 0)
+                return r;
```
Could you add a brief comment at the beginning of `on_stream_io_impl()` to mention that it returns negative errno on critical error, 0 when recoverable error or the stream handling is finished. and 1 for possibly remaining stream exists.

yuwata commented on src/resolve/resolved-dns-stream.c:
```diff
@@ -441,6 +439,45 @@ static int on_stream_io(sd_event_source *es, int fd, uint32_t revents, void *use
+                return r;
+
+#if ENABLE_DNS_OVER_TLS
+        if (s->encrypted) {
```
Let's reduce the indent:
```
if (!s->encrypted)
        return 0;
```

yuwata commented on src/resolve/resolved-dns-stream.c:
```diff
@@ -441,6 +439,45 @@ static int on_stream_io(sd_event_source *es, int fd, uint32_t revents, void *use
+        if (s->encrypted) {
+                uint32_t events;
+
+                r = sd_event_source_get_io_events(s->io_event_source, &events);
```
Why this is called?? Why do not you reuse `revents` here?

joanbm replied:
From `man 2 poll`, `revents` are the returned events (i.e. which events we received to process), `events` are the requested events (i.e. which events we could process next).

The point of checking whether `FLAGS_SET(events, EPOLLIN)` is to know whether the stream still wants to read more data. It would be wrong to return our "pseudo-`EPOLLIN` event " on buffered TLS data if the stream is not even expecting to receive further data available events.

Note that in this case `events` is just what was set by the last call to `dns_stream_update_io`. For example, looking at the condition on `dns_stream_update_io`, when we already have a full DNS packet stored on `s->read_packet`, we don't want to process any further data, even if it's available.

PS: BTW, something unrelated to this issue which I noticed when developing the test, is that `dns_stream_take_read_packet`doesn't call `dns_stream_update_io`. What this means is that unless you process the packets on the (supposedly optional) `on_packet` callback, taking the packet doesn't set the `EPOLLIN` flag again, so no further packets will be processed even if they are available... which seems wrong. But all callers use the `on_packet` callback, so this is just a theoretical issue for now.

yuwata replied:
That makes sense for me. Thanks.

yuwata replied:
> PS: BTW, something unrelated to this issue which I noticed when developing the test, is that `dns_stream_take_read_packet`doesn't call `dns_stream_update_io`. What this means is that unless you process the packets on the (supposedly optional) `on_packet` callback, taking the packet doesn't set the `EPOLLIN` flag again, so no further packets will be processed even if they are available... which seems wrong. But all callers use the `on_packet` callback, so this is just a theoretical issue for now.

I cannot catch the issue... There is `dns_stream_update_io()` after `on_packet()`. Is it not enough?

joanbm replied:
Imagine I had written the loop of line 300 in `src/resolve/test-resolved-stream.c ` like this:
```c
        stream->on_packet = NULL;
        while (n_received_packets != 2) {
                assert_se(sd_event_run(event, EVENT_TIMEOUT_USEC) >= 1);
                DnsPacket *p = dns_stream_take_read_packet(stream);
                if (p)
                        received_packets[n_received_packets++] = p;
        }
```

IMO this ought to work, but it actually fails for both TCP and TLS, reading only a single packet, then timing out.

Why? Because when I call `sd_event_run`, `s->read_packet` is filled. When `dns_stream_update_io` is called after reading the packet in `on_stream_io_impl`, the `EPOLLIN` event flag is unset. Later I call `dns_stream_take_read_packet` so `s->read_packet` is free again, but no one sets the `EPOLLIN` event flag again, so no further packets are read and the test times out.

But all current usages call `dns_stream_take_read_packet` *inside* the `on_packet` callback rather than outside, so the `EPOLLIN` event flag is never unset, and everything works.

yuwata replied:
Thank you for the explanation! Let's add a brief comment in `dns_stream_take_read_packet()`.

yuwata replied:
BTW, while I try to understand your comment previously, I found a potential issue, and filed as #22266. If you'd like, PTAL.

yuwata replied:
> Thank you for the explanation! Let's add a brief comment in `dns_stream_take_read_packet()`.

#22274.

mrc0mmand commented:
>    * Fix CI: Skip the unit test if the 'openssl' command is not available at runtime. @mrc0mmand not sure if you have any better way to go about this. I don't like to use OpenSSL or GnuTLS as a library in the test, I prefer something like openssl/socat/stunnel/etc. since it's simpler, and harder to get wrong.

I definitely agree, so using the `openssl` utility here is fine by me.

> PS: Sorry for launching so many CI runs, I kept finding small things to fix.

Don't worry about it, most of the CIs automagically cancel the previous runs on force push.

yuwata reviewed (approved):
LGTM. Thank you!

(I dropped doubled empty line. And re-ordered the commits, as the test does not pass without the fix.)

yuwata commented:
BTW, I noticed that this extremely improves performance (~7x faster :+1: ):

With
```
$ resolvectl status enp0s20f0u2 
Link 4 (enp0s20f0u2)
    Current Scopes: DNS LLMNR/IPv4 LLMNR/IPv6
         Protocols: +DefaultRoute +LLMNR -mDNS +DNSOverTLS DNSSEC=yes/supported
Current DNS Server: 1.1.1.1
       DNS Servers: 1.1.1.1 1.0.0.1
```

Before:
```
$ resolvectl --cache=no query www.youtube.com
www.youtube.com: 172.217.31.142                -- link: enp0s20f0u2
                 172.217.161.46                -- link: enp0s20f0u2
                 142.250.196.142               -- link: enp0s20f0u2
                 172.217.175.14                -- link: enp0s20f0u2
                 172.217.175.46                -- link: enp0s20f0u2
                 172.217.175.78                -- link: enp0s20f0u2
                 172.217.175.110               -- link: enp0s20f0u2
                 172.217.31.174                -- link: enp0s20f0u2
                 216.58.220.142                -- link: enp0s20f0u2
                 172.217.161.78                -- link: enp0s20f0u2
                 172.217.175.238               -- link: enp0s20f0u2
                 216.58.220.110                -- link: enp0s20f0u2
                 172.217.174.110               -- link: enp0s20f0u2
                 142.251.42.142                -- link: enp0s20f0u2
                 142.250.196.110               -- link: enp0s20f0u2
                 142.251.42.174                -- link: enp0s20f0u2
                 2404:6800:4004:813::200e      -- link: enp0s20f0u2
                 2404:6800:4004:80b::200e      -- link: enp0s20f0u2
                 2404:6800:4004:808::200e      -- link: enp0s20f0u2
                 2404:6800:4004:81c::200e      -- link: enp0s20f0u2
                 (youtube-ui.l.google.com)

-- Information acquired via protocol DNS in 635.8ms.
-- Data is authenticated: no; Data was acquired via local or encrypted transport: yes
-- Data from: network
```
```
$ resolvectl --cache=no query www.youtube.com
www.youtube.com: 2404:6800:4004:826::200e      -- link: enp0s20f0u2
                 2404:6800:4004:80f::200e      -- link: enp0s20f0u2
                 2404:6800:4004:801::200e      -- link: enp0s20f0u2
                 2404:6800:4004:827::200e      -- link: enp0s20f0u2
                 142.250.207.46                -- link: enp0s20f0u2
                 216.58.220.110                -- link: enp0s20f0u2
                 142.250.196.110               -- link: enp0s20f0u2
                 172.217.25.78                 -- link: enp0s20f0u2
                 216.58.197.238                -- link: enp0s20f0u2
                 142.250.199.110               -- link: enp0s20f0u2
                 142.251.42.174                -- link: enp0s20f0u2
                 172.217.161.78                -- link: enp0s20f0u2
                 216.58.197.206                -- link: enp0s20f0u2
                 142.250.207.14                -- link: enp0s20f0u2
                 172.217.25.110                -- link: enp0s20f0u2
                 172.217.31.142                -- link: enp0s20f0u2
                 142.251.42.142                -- link: enp0s20f0u2
                 172.217.26.238                -- link: enp0s20f0u2
                 142.250.196.142               -- link: enp0s20f0u2
                 142.251.42.206                -- link: enp0s20f0u2
                 (youtube-ui.l.google.com)

-- Information acquired via protocol DNS in 90.9ms.
-- Data is authenticated: no; Data was acquired via local or encrypted transport: no
-- Data from: network
```

foutrelis commented:
Does this fix meet the criteria for a stable backport?

Context: I'm using systemd 250.4 and noticed page loads being delayed in Chromium shortly after enabling `DNSOverTLS`. After applying 2aaf6bb6e99b0f2bd73e0c49bef9e11a2844bf1a on top of v250.4, `resolvectl query google.com --cache=no` no longer hangs. 🐭

foutrelis commented:
Does this fix meet the criteria for a stable backport?

Context: I'm using systemd 250.4 and noticed page loads being delayed in Chromium shortly after enabling `DNSOverTLS`. After applying 2aaf6bb6e99b0f2bd73e0c49bef9e11a2844bf1a on top of v250.4, `resolvectl query google.com --cache=no` no longer hangs. 🐭
